File: src/main/java/com/kapilkoju/autopos/domain/Vendor.java
package com.kapilkoju.autopos.domain;

import com.fasterxml.jackson.annotation.JsonIgnore;

import javax.persistence.*;
import javax.validation.constraints.*;
import java.io.Serializable;
import java.util.HashSet;
import java.util.Set;
import java.util.Objects;

/**
 * A Vendor.
 */
@Entity
@Table(name = "vendor")
public class Vendor implements Serializable {

    private static final long serialVersionUID = 1L;

    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "sequenceGenerator")
    @SequenceGenerator(name = "sequenceGenerator")
    private Long id;

    @NotNull
    @Size(min = 2, max = 100)
    @Column(name = "name", length = 100, nullable = false)
    private String name;

    @Size(max = 250)
    @Column(name = "remarks", length = 250)
    private String remarks;

    @OneToMany(mappedBy = "vendor")
    @JsonIgnore
    private Set<Phone> phones = new HashSet<>();

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public Vendor name(String name) {
        this.name = name;
        return this;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getRemarks() {
        return remarks;
    }

    public Vendor remarks(String remarks) {
        this.remarks = remarks;
        return this;
    }

    public void setRemarks(String remarks) {
        this.remarks = remarks;
    }

    public Set<Phone> getPhones() {
        return phones;
    }

    public Vendor phones(Set<Phone> phones) {
        this.phones = phones;
        return this;
    }

    public Vendor addPhone(Phone phone) {
        this.phones.add(phone);
        return this;
    }

    public Vendor removePhone(Phone phone) {
        this.phones.remove(phone);
        return this;
    }

    public void setPhones(Set<Phone> phones) {
        this.phones = phones;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Vendor vendor = (Vendor) o;
        if (vendor.id == null || id == null) {
            return false;
        }
        return Objects.equals(id, vendor.id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    @Override
    public String toString() {
        return "Vendor{" +
            "id=" + id +
            ", name='" + name + "'" +
            ", remarks='" + remarks + "'" +
            '}';
    }
}

File: kotlang/src/main/kotlin/com/kapilkoju/autopos/party/domain/Vendor.kt
package com.kapilkoju.autopos.party.domain

import com.kapilkoju.autopos.kernel.domain.AuditableEntity

import javax.persistence.*
import javax.validation.constraints.NotNull
import javax.validation.constraints.Size


@Entity
@Table(name = "vendor")
@NamedEntityGraph(name = Vendor.Graph.detail,
        attributeNodes = [NamedAttributeNode("phoneNumbers")])
data class Vendor(

        @NotNull
        @Size(min = 2, max = 100)
        @Column(name = "name", length = 100, nullable = false, unique = true)

        val name: String,

        @ElementCollection
        @CollectionTable(name = "vendor_phone_numbers")
        @Column(name = "phone_number")
        val phoneNumbers: List<String> = ArrayList(),

        @Size(max = 250)
        @Column(name = "remarks", length = 250)

        val remarks: String?

) : AuditableEntity() {

    companion object Graph {
        const val detail = "Vendor.detail"
    }

}


